Rework resource guide (#3485)
tauri-apps/tauri-docs: src/content/docs/develop/resources.mdx

```json title=tauri.conf.json5
{
  "bundle": {
    "resources": {
      // `textfile.txt` will be placed to `$RESOURCE/resources/textfile.txt`
      "/absolute/path/to/textfile.txt": "resources/textfile.txt",

      // `jsonfile.json` will be placed to `$RESOURCE/resources/jsonfile.json`
      "relative/path/to/jsonfile.json": "resources/jsonfile.json",

      // Copy the entire directory to `$RESOURCE`, preserving the original structures,
      // the target is "" which means it will be placed directly in the resource directory `$RESOURCE`, for example:
      //   - `resources/file.txt`                -> `$RESOURCE/file.txt`
      //   - `resources/some-folder/config.json` -> `$RESOURCE/some-folder/config.json`
      "resources/": "",

      // When using glob patterns, the behavior is different from the list one,
      // all the matching files will be placed to the target directory without preserving the original file structures
      // for example:
      //   - `docs/index.md`         -> `$RESOURCE/website-docs/index.md`
      //   - `docs/plugins/setup.md` -> `$RESOURCE/website-docs/setup.md`
      "docs/**/*md": "website-docs/"
    }
  }
}
```

</TabItem>
</Tabs>

> To learn about where does `$RESOURCE` resolve to on each platforms, see the documentation of [`resource_dir`]

Reviewer: ```suggestion
To learn about where `$RESOURCE` resolves to on each platforms, see the documentation of [`resource_dir`]
```

To fine control where the files will be copied to, use a map instead:

```json title=tauri.conf.json
```json title=tauri.conf.json5
{
  "bundle": {
    "resources": {
      // `textfile.txt` will be placed at `$RESOURCE/resources/textfile.txt`
      "/absolute/path/to/textfile.txt": "resources/textfile.txt",

      // `jsonfile.json` will be placed at `$RESOURCE/resources/jsonfile.json`
      "relative/path/to/jsonfile.json": "resources/jsonfile.json",
      "resources/**/*": "resources/"

      // We remove the pattern from the left side if it matches the right side,
      // for example:
      //   - `resources/file.text` -> `$RESOURCE/resources/file.text`
      //   - `resources/file.text` -> `$RESOURCE/resources/file.text`
      "resources/**/*": "resources/",

      // In this case, the pattern doesn't match, so we preserve the original structures:
      //   - `resources/file.text`        -> `$RESOURCE/other-resources/resources/file.text`
      //   - `resources/folder/file.text` -> `$RESOURCE/other-resources/resources/folder/file.text`
      "resources/**/*": "other-resources/",
    }
  }
}
```

> :::note

Reviewer: Unless this is not true anymore, please do not remove it as it was added after users had issues with it.

(The diff is super hard to read on mobile so i can't tell if it was integrated somewhere else)

Author: Since we already have a note about `fs:allow-resource-read-recursive` below and we have explained where the files will be going, I think it's clear enough that when doing the scope thing, a recursive one should be used, and if people still can't get it, I think we should spend a bit on the scope documentations (e.g. the fs scope capability's description) instead

The old notice here feels a bit out of place, since we didn't mention anything permission related at that point, unlike our v1 doc page https://v1.tauri.app/v1/guides/building/resources/

Reviewer 2: maybe add a permissions section here instead? i feel like removing it makes it too easy for people to misconfigure the capabilities

Author: I think a separate section makes sense, I could add that tomorrow

  let resource_path = handle.path().resolve("lang/de.json", BaseDirectory::Resource)?;
}
```

</TabItem>
<TabItem label="JavaScript">

To resolve the path in JavaScript, use [`resolveResource`]:

```javascript
import { resolveResource } from '@tauri-apps/api/path';
const resourcePath = await resolveResource('lang/de.json');
```

</TabItem>
</Tabs>

### Path syntax

The path in the API calls can be either a normal relative path like `folder/json_file.json` that resolves to `$RESOURCE/folder/json_file.json`,
or a paths like `../relative/folder/toml_file.toml` that resolves to `$RESOURCE/_up_/relative/folder/toml_file.toml`,
these APIs use the same rules as you write `tauri.conf.json > bundle > resources`, for example:

```json title=tauri.conf.json
{
  "bundle": {
    "resources": ["folder/json_file.json", "../relative/folder/toml_file.toml"]
  }
}
```

```rust
let json_path = app.path().resolve("folder/json_file.json", BaseDirectory::Resource)?;
let toml_path = app.path().resolve("../relative/folder/toml_file.toml", BaseDirectory::Resource)?;
```

### Android

Currently the resources are stored in the APK as assets so the return value of those APIs are not normal file system paths,

> we use a special URI prefix `asset://localhost/` here that can be used with the [`plugin-fs`],

Reviewer: ```suggestion
we use a special URI prefix `asset://localhost/` here that can be used with the [`fs` plugin],
```

> we use a special URI prefix `asset://localhost/` here that can be used with the [`plugin-fs`],
> with that, you can read the files through [`FsExt::fs`] like this:

Reviewer: ```suggestion
we use a special URI prefix `asset://localhost/` here that can be used with the [`fs` plugin].
With that plugin, you can read the files through [`FsExt::fs`] like this:
```

```rust
let resource_path = app.path().resolve("lang/de.json", BaseDirectory::Resource).unwrap();
let json = app.fs().read_to_string(&resource_path);
```

> If you want or must need the resource files to be on a real file system, copy the contents out manually through the [`plugin-fs`]

Reviewer: ```suggestion
If you want or must have the resource files to be on a real file system, copy the contents out manually through the [`fs` plugin]
```

fn hello(handle: tauri::AppHandle) -> String {
    let resource_path = handle.path().resolve("lang/de.json", BaseDirectory::Resource)?;

    let json = std::fs::read_to_string(&resource_path).unwrap();
    // Or when dealing with Android, use the file system plugin instead
    // let json = handle.fs().read_to_string(&resource_path);

    let lang_de: serde_json::Value = serde_json::from_str(json).unwrap();

    lang_de.get("hello").unwrap()
}
```

### JavaScript

> For the JavaScript side, you can either use a command like the one above and call it through `await invoke('hello')` or access the files using the [`plugin-fs`]

Reviewer: ```suggestion
For the JavaScript side, you can either use a command like the one above and call it through `await invoke('hello')` or access the files using the [`fs` plugin].
```
not sure if this suggestion works or if i missed some that don't show up in the diff.

Author: Seems like you've cached everything, GitHub's website doesn't let me to commit the suggestions though, did it locally

console.log(langDe.hello); // This will print 'Guten Tag!' to the devtools console
```

## Permissions

Since we replace `../` to `_up_` in relative paths and the root to `_root_` in abosolute paths when using a list,
those files will be in sub folders inside the resource directory,
to allow those paths in Tauri's [permission system](/security/capabilities/),
use `$RESOURCE/**/*` to allow recursive access to those files

### Examples

With a file bundled like this:

```json title=tauri.conf.json
{
  "bundle": {
    "resources": ["../relative/path/to/jsonfile.json"]
  }
}
```

> To use it with the [`plugin-fs`]:

Reviewer: ```suggestion
To use it with the [`fs` plugin]:
```

```json title=src-tauri/capabilities/default.json ins={8-15}
{
  "$schema": "../gen/schemas/desktop-schema.json",
  "identifier": "main-capability",
  "description": "Capability for the main window",
  "windows": ["main"],
  "permissions": [
    "core:default",
    "fs:allow-stat",
    "fs:allow-read-text-file",
    "fs:allow-resource-read-recursive",
    {
      "identifier": "fs:scope",
      "allow": ["$RESOURCE/**/*"],
      "deny": ["$RESOURCE/secret.txt"]
    }
  ]
}
```

> To use it with the [`plugin-opener`]:

Reviewer: ```suggestion
To use it with the [`opener` plugin]:
```

```json title=src-tauri/capabilities/default.json ins={8-15}
{
  "$schema": "../gen/schemas/desktop-schema.json",
  "identifier": "main-capability",
  "description": "Capability for the main window",
  "windows": ["main"],
  "permissions": [
    "core:default",
    {
      "identifier": "opener:allow-open-path",
      "allow": [
        {
          "path": "$RESOURCE/**/*"
        }
      ]
    }
  ]
}
```

[`resource_dir`]: https://docs.rs/tauri/latest/tauri/path/struct.PathResolver.html#method.resource_dir
[`pathresolver`]: https://docs.rs/tauri/latest/tauri/path/struct.PathResolver.html
[`PathResolver::resolve`]: https://docs.rs/tauri/latest/tauri/path/struct.PathResolver.html#method.resolve
[`resolveResource`]: https://tauri.app/reference/javascript/api/namespacepath/#resolveresource
[`app`]: https://docs.rs/tauri/latest/tauri/struct.App.html
[`apphandle`]: https://docs.rs/tauri/latest/tauri/struct.AppHandle.html

> [`plugin-fs`]: /plugin/file-system/

Reviewer: ```suggestion
[`fs` plugin]: /plugin/file-system/
```

[`FsExt::fs`]: https://docs.rs/tauri-plugin-fs/latest/tauri_plugin_fs/trait.FsExt.html#tymethod.fs
[basic setup]: /plugin/file-system/#setup
[Scope Permissions]: /plugin/file-system/#scopes
[scopes]: /plugin/file-system/#scopes

> [`plugin-opener`]: /plugin/opener/

Author: ```suggestion
[`opener` plugin]: /plugin/opener/
```
